### PR TITLE
Fix view filter update and deletion propagation

### LINKED ISSUE #731: Investigate impact of fields deletion and select option updates/deletion on viewFields/viewFilters, etc.
It seems like there are some issues when we modify the data model and it's not properly propagated to the view.

For example, for options we shouldn't hardcode the option value in the view filter but instead use the option ID which will work out better with umdates.

### PR DESCRIPTION
**Pull request by prastoin**
# Introduction

Diff description: ~500 tests and +500 additions

close https://github.com/twentyhq/core-team-issues/issues/731

## What has been done here
In a nutshell on a field metadata type ( `SELECT MULTI_SELECT` ) update, we will be browsing all `ViewFilters` in a post hook searching for some referencing related updated `fieldMetadata` select. In order to update or delete the `viewFilter` depending on the associated mutations.

## How to test:
- Add FieldMetadata `SELECT | MULTI_SELECT` to an existing or a new `objectMetadata`
- Create a filtered view on created `fieldMetadata` with any options you would like
- Remove some options ( in the best of the world some that are selected by the filter ) from the `fieldMetadata` settings page
- Go back to the filtered view, removed or updated options should have been hydrated in the `displayValue` and the filtered data should make sense

## All filtered options are deleted edge case
If an update implies that a viewFilter does not have any existing related options anymore, then we remove the viewFilter

## Testing
```sh 
PASS  test/integration/metadata/suites/field-metadata/update-one-field-metadata-related-record.integration-spec.ts (27 s)
  update-one-field-metadata-related-record
    SELECT
      ✓ should delete related view filter if all select field options got deleted (2799 ms)
      ✓ should update related multi selected options view filter (1244 ms)
      ✓ should update related solo selected option view filter (1235 ms)
      ✓ should handle partial deletion of selected options in view filter (1210 ms)
      ✓ should handle reordering of options while maintaining view filter values (1487 ms)
      ✓ should handle no changes update of options while maintaining existing view filter values (1174 ms)
      ✓ should handle adding new options while maintaining existing view filter (1174 ms)
      ✓ should update display value with options label if less than 3 options are selected (1249 ms)
      ✓ should throw error if view filter value is not a stringified JSON array (1300 ms)
    MULTI_SELECT
      ✓ should delete related view filter if all select field options got deleted (1127 ms)
      ✓ should update related multi selected options view filter (1215 ms)
      ✓ should update related solo selected option view filter (1404 ms)
      ✓ should handle partial deletion of selected options in view filter (1936 ms)
      ✓ should handle reordering of options while maintaining view filter values (1261 ms)
      ✓ should handle no changes update of options while maintaining existing view filter values (1831 ms)
      ✓ should handle adding new options while maintaining existing view filter (1610 ms)
      ✓ should update display value with options label if less than 3 options are selected (1889 ms)
      ✓ should throw error if view filter value is not a stringified JSON array (1365 ms)

Test Suites: 1 passed, 1 total
Tests:       18 passed, 18 total
Snapshots:   18 passed, 18 total
Time:        27.039 s
```
## Out of scope
- We should handle ViewFilter validation when extracting its definition from the metadata https://github.com/twentyhq/core-team-issues/issues/1009

## Concerns
- Are we able through the api to update an RATING fieldMetadata ? ( if yes than that's an issue and we should handle RATING the same way than for SELECT and MULTI_SELECT )
- It's not possible to group a view from a MULTI_SELECT field

The above points create a double nor a triple "lecture" to the post hook effect:
- ViewGroup -> only SELECT
- VIewFilter -> only SELECT || MULTI_SELECT
- Rating nothing
I think we should determine the scope of all of that

### CODE REVIEW

**Comment by prastoin** [comment]
discovered that the viewGroup related on update ( not delete or create option ) is broken, fixing

File: packages/twenty-server/test/integration/metadata/types/common-response-body.type.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { BaseGraphQLError } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';
+
+export type CommonResponseBody<T> = Promise<{
+  data: T;
+  errors: BaseGraphQLError[];
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/metadata/types/common-response-body.type.ts:5]: You're right would be better but from now leaving it like that about to rework few things in the integrations tests tooling

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 2]
```diff
@@ -19,24 +27,37 @@ type Differences<T> = {
...
+    // TODO legacy should support multi-select and rating ?
     if (
       !isSelectFieldMetadataType(newFieldMetadata.type) ||
       !isSelectFieldMetadataType(oldFieldMetadata.type)
     ) {
       return;
     }
-
-    const views = await this.getFieldMetadataViews(newFieldMetadata);
+    const views = await this.getFieldMetadataViewEntity(
```
Comment (guillim) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:57]: getFieldMetadataViewEntities would be clearer, wouldn't it ?

Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:57]: The function only returns one entity

File: packages/twenty-server/test/integration/metadata/suites/field-metadata/update-one-field-metadata-related-record.integration-spec.ts [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        );
+
+        expect(findResponse).toMatchSnapshot({
+          id: expect.any(String),
+        });
+      },
+    );
+
+    // Note these test exists only because we do not validate the view filter value on creation/update
+    // Should be removed after https://github.com/twentyhq/core-team-issues/issues/1009 completion
```
Comment (prastoin) [inline packages/twenty-server/test/integration/metadata/suites/field-metadata/update-one-field-metadata-related-record.integration-spec.ts:301]: ***Request_changes:*** Broken link

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts (hunk 2 above)
Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:57]: In my mind the `entity` is either `viewFilter` or `viewGroup` and It the methods returns the occurrences of this entity 🤔

Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:57]: I kept the previously named `views` const but indeed it's not accurate, could be renamed to `viewGroups` or`groups`

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 4]
```diff
@@ -159,8 +280,9 @@ export class FieldMetadataRelatedRecordsService {
     return differences;
   }
 
-  private async getFieldMetadataViews(
-    fieldMetadata: FieldMetadataEntity,
+  private async getFieldMetadataViewEntity(
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:283]: just my opinion but I think I found `getFieldMetadataViews` more readable
at least plural seems important to keep here i would say.
Also I find entity arg a bit misleading as it's not the same entity as in the name
Maybe `private async getFieldMetadataView({fieldMetadata, entityFilter}} `? the entity from args is actually used as a filter to retrieve

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 5]
```diff
@@ -100,8 +121,108 @@ export class FieldMetadataRelatedRecordsService {
...
+  public async updateRelatedViewFilters(
+    oldFieldMetadata: SelectFieldMetadataEntity,
+    newFieldMetadata: SelectFieldMetadataEntity,
+  ): Promise<void> {
+    const filters = await this.getFieldMetadataViewEntity(
+      newFieldMetadata,
+      'viewFilters',
+    );
+
+    const { updated: updatedFieldMetadata, deleted: deletedFieldMetadata } =
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:143]: I really think we should update the name to 
```suggestion
    const { updated: updatedOptions, deleted: deletedOptions } =
```
or 
```suggestion
    const { updated: updatedFieldMetadataOptions, deleted: deletedFieldMetadataOptions } =
```

because updatedFieldMetadata makes me expect a FieldMetadataEntity

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 6]
```diff
@@ -100,8 +121,108 @@ export class FieldMetadataRelatedRecordsService {
...
+    }
+
+    return newViewFilterOptions.map((option) => option.label).join(', ');
+  }
+
+  public async updateRelatedViewFilters(
+    oldFieldMetadata: SelectFieldMetadataEntity,
+    newFieldMetadata: SelectFieldMetadataEntity,
+  ): Promise<void> {
+    const filters = await this.getFieldMetadataViewEntity(
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:138]: those aren't `filters` but `views` right?

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 7]
```diff
@@ -100,8 +121,108 @@ export class FieldMetadataRelatedRecordsService {
...
+        continue;
+      }
+
+      for (const viewFilter of filter.viewFilters) {
+        const viewFilterValue = parseJson<string[]>(viewFilter.value);
+
+        if (!isDefined(viewFilterValue) || !Array.isArray(viewFilterValue)) {
+          throw new FieldMetadataException(
+            `Invalid view filter value for view filter ${viewFilter.id}`,
+            FieldMetadataExceptionCode.INVALID_FIELD_INPUT,
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:173]: I don't think it's the right error here: the viewFilterValue does not come from the user input who is currently editing a field's options, not the already existing viewFilter values. 
I think this is an INTERNAL_SERVER_ERROR rather - `Unexpected invalid view filter value` maybe.

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 8]
```diff
@@ -100,8 +121,108 @@ export class FieldMetadataRelatedRecordsService {
...
+
+        const viewFilterOptions = viewFilterValue
+          .map((value) =>
+            oldFieldMetadata.options.find((option) => option.value === value),
+          )
+          .filter(isDefined);
+
+        const afterDeleteViewFilterOptions = viewFilterOptions.filter(
+          (viewFilterOption) =>
+            !deletedFieldMetadata.some(
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:185]: `deletedFieldMetadataOptions`

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 9]
```diff
@@ -100,8 +121,108 @@ export class FieldMetadataRelatedRecordsService {
...
+              (option) => option.value === viewFilterOption.value,
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:186]: can we do the deletion based on the id rather ? In case the user deletes an option and re-creates or updates an existing one with the same value as the one they just deleted (I guess that could happen)

Comment (ijreilly) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:186]: maybe it's a usecase we should test

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 10]
```diff
@@ -100,8 +121,108 @@ export class FieldMetadataRelatedRecordsService {
...
+            ),
+        );
+
+        if (afterDeleteViewFilterOptions.length === 0) {
+          await viewFilterRepository.delete({ id: viewFilter.id });
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:191]: have you checked the implications on viewFilterGroup? 
Maybe we should have a dedicated viewFilterService to delete a viewFilterGroup when it comes to having 0 viewFilter left

(i don't know viewFilterGroup well, maybe that does not make sense)

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts [hunk 11]
```diff
@@ -100,8 +121,108 @@ export class FieldMetadataRelatedRecordsService {
...
+          continue;
+        }
+
+        const afterUpdateAndDeleteViewFilterOptions =
+          afterDeleteViewFilterOptions.map((viewFilterOption) => {
+            const updatedOption = updatedFieldMetadata.find(
+              ({ old }) => viewFilterOption.value === old.value,
```
Comment (ijreilly) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:198]: can we use option ids here instead?

File: packages/twenty-server/test/integration/graphql/utils/create-one-operation.util.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import { PerformMetadataQueryParams } from 'test/integration/metadata/types/perform-metadata-query.type';
+import { warnIfNoErrorButExpectedToFail } from 'test/integration/metadata/utils/warn-if-no-error-but-expected-to-fail.util';
+import { capitalize } from 'twenty-shared/utils';
+
+import { ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';
+
+type CreateOneOperationArgs<T> = PerformMetadataQueryParams<T> & {
+  objectMetadataSingularName: string;
+};
+export const createOneOperation = async <T = object>({
```
Comment (ijreilly) [inline packages/twenty-server/test/integration/graphql/utils/create-one-operation.util.ts:13]: great !

File: packages/twenty-server/test/integration/metadata/suites/field-metadata/__snapshots__/update-one-field-metadata-related-record.integration-spec.ts.snap [hunk 13]
```diff
@@ -0,0 +1,157 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`update-one-field-metadata-related-record MULTI_SELECT should delete related view filter if all select field options got deleted 1`] = `
```
Comment (ijreilly) [inline packages/twenty-server/test/integration/metadata/suites/field-metadata/__snapshots__/update-one-field-metadata-related-record.integration-spec.ts.snap:3]: looks cool, how can we see it ? or is this just a test to run? 
Should be shared with the team !

**Review by lucasbordeau — COMMENTED** [review]
LGTM for frontend nice PR

**Review by ijreilly — CHANGES_REQUESTED** [review]
Some notes

1/ Great for the tests, that's the right mindset it is exactly what we need to do now :)

2/ FE is not updated until the page is refreshed
Capture d’écran 2025-05-27 à 11 38 06

3/ Unexpected behaviour if I delete my option "Meeting" then update another option with the name "Meeting" (was maybe already there given how differences are computed but worth fixing now imo): the new "Meeting" is still in the options but not showing on the views anymore (i think)

4/ Some names are misleading

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts (hunk 4 above)
Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:283]: As @guillim recommended about to suggestion a new name

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts (hunk 8 above)
Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:185]: ?

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts (hunk 10 above)
Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:191]: Yes, unless I'm mistaken that's the same for the `filterGroup` implem that consumes the `viewGroupRepository` directly.
Mhmm yes that would make sense, but I'm not sure of the DevX. Calling an update with empty options resulting in deletion ? I think this responsability is more to the caller rather than the repo :thinking:

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts (hunk 9 above)
Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:186]: That's the limitation of the current implementation based on the `value` rather than the `id`.
We've been discussing migrating to the id pattern for more info please see https://github.com/twentyhq/core-team-issues/issues/731#issuecomment-2880775314

Comment (prastoin) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/services/field-metadata-related-records.service.ts:186]: > can we do the deletion based on the id rather ? In case the user deletes an option and re-creates or updates an existing one with the same value as the one they just deleted (I guess that could happen)

When updating options you don't really delete anything explicitly, you override. Mutating an id implies an options deletion. Unless I'm mistaken in some way this is already covered by the update-create-field-metdata-select test suite

**Review by ijreilly — APPROVED** [review]
ok for me, i think we should do an extensive QA on main when this is merged
